packages/back-end/src/services/jobLifecycle.ts

```diff
@@ -0,0 +1,51 @@
+import { Job, JobAttributesData } from "agenda";
+import { logger } from "back-end/src/util/logger";
+import { MAX_QUERY_TIMEOUT_MS } from "back-end/src/util/secrets";
+
+const TOUCH_INTERVAL_MS = 3 * 60 * 1000;
+const JOB_TIMEOUT_MS = MAX_QUERY_TIMEOUT_MS + 1000; // Allow some buffer for the query client to close properly after it times out
+
+// As some functions may take a while to run, this will update the job's lock if the process is still running
```

Comment:
We could add to the end of this comment some more comments:
```
.
//This prevents the lockLifetime being reached as long as the job is running, and hence stops other servers from picking up the job.
//This also adds a timeout which allows the job to keep running but marks it as failed, which frees up the "slot" for another job to run, in case the defaultLockLimit is reached, and also prevents other jobs from picking it up unless they have retry logic.